### PR TITLE
fix: fall back to system memory if cgroup exceeds the available memory

### PR DESCRIPTION
**Pull request by kruskall**
## Motivation/summary
I couldn't find any issue in the cgroup reading logic that would explain the bug in https://github.com/elastic/apm-server/pull/9868

On the other hand, it looks like `9223372036854771712` is the default value for cgroup v1 for unlimited memory (no limit).

I was able to reproduce the content of the limit_in_bytes file by creating a fedora 30 instance in aws and running a docker container inside of it. Then reading the memory limit inside the container:

```
# cat /sys/fs/cgroup/memory/memory.limit_in_bytes 
9223372036854771712
```

Note: fedora 30 is important as later versions are using cgroup v2

The PR is adding a fall back to system memory if cgroup returns a value that exceeds the total memory.

## Checklist

- [ ] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)
- [ ] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)
- [ ] Documentation has been updated

For functional changes, consider:
- Is it observable through the addition of either **logging** or **metrics**?
- Is its use being published in **telemetry** to enable product improvement?
- Have system tests been added to avoid regression?

### CODE REVIEW

File: internal/beater/beater.go [hunk 1]
```diff
@@ -189,15 +189,19 @@ func (s *Runner) Run(ctx context.Context) error {
...
+	} else {
+		if memLimit <= 0 {
+			s.logger.Info("no cgroups detected, falling back to total system memory")
 			// If no cgroup limit is set, only return 50% of the total memory.
 			// to have a margin of safety for other processes.
 			memLimit = float64(limit) / 1024 / 1024 / 1024 * 0.5
 		}
+		if memLimit > float64(limit) {
+			s.logger.Info("cgroup memory limit exceed available memory, falling back to total system memory")
+			memLimit = float64(limit)
```
Comment (lahsivjar) [inline internal/beater/beater.go:203]: I think the case for keeping the memory limit to be a fraction of total memory should hold for this case also.

**Review by lahsivjar — COMMENTED** [review]
LGTM other than the final value of the limit.

File: internal/beater/beater.go (hunk 1 above)
Comment (kruskall) [reply internal/beater/beater.go]: Agreed! Updated :+1:

**Review by lahsivjar — APPROVED** [review]
LGTM!

**Comment by lahsivjar** [comment]
@simitt Should we backport this to `8.6`?

**Review by simitt — APPROVED** [review]
Great work @kruskall in reproducing and finding the root cause 🚀

**Comment by simitt** [comment]
>@simitt Should we backport this to 8.6?

yes, it would introduce a known bug otherwise, and I consider it low risk to backport. @kruskall can you take care of it please?

**Pull request merged**